File: src/main/java/com/amaze/filemanager/ui/views/preference/PathSwitchPreference.java
package com.amaze.filemanager.ui.views.preference;

import android.content.Context;
import android.preference.Preference;
import android.support.annotation.IdRes;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Switch;

import com.amaze.filemanager.R;

/**
 * @author Emmanuel
 *         on 17/4/2017, at 22:22.
 */

public class PathSwitchPreference extends Preference {

    public static final int EDIT = 0, SWITCH = 1, DELETE = 2;

    private int lastItemClicked = -1;
    private Switch switchView;
    private View.OnClickListener switchListener;
    /**
     * shouldEnable is the same thing as enabled, but is used before super.onBindView(view) has been called
     * enabled is the current state of this Preference (check this.updateSwitch())
     */
    private boolean shouldEnable = true, enabled = true;

    public PathSwitchPreference(Context context) {
        super(context);
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        setWidgetLayoutResource(R.layout.namepathswich_preference);
        return super.onCreateView(parent);
    }

    @Override
    protected void onBindView(View view) {
        setListener(view, R.id.edit, EDIT);
        switchListener = setListener(view, R.id.switch_button, SWITCH);
        setListener(view, R.id.delete, DELETE);

        super.onBindView(view);//Keep this before things that need changing what's on screen

        switchView = (Switch) view.findViewById(R.id.switch_button);

        switchView.setChecked(shouldEnable);
        updateSwitch(view);
    }

    public void setChecked(boolean checked) {
        if(switchView != null) {
            switchView.setChecked(checked);
            switchListener.onClick(switchView);
        } else shouldEnable = checked;
    }

    public boolean isChecked() {
        return enabled;
    }

    public int getLastItemClicked() {
        return lastItemClicked;
    }

    private View.OnClickListener setListener(final View v, @IdRes int id, final int elem) {
        final PathSwitchPreference t = this;

        View.OnClickListener l = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                lastItemClicked = elem;

                if(lastItemClicked == SWITCH) {
                    updateSwitch(v);
                }

                if(enabled || lastItemClicked != EDIT) {
                    getOnPreferenceClickListener().onPreferenceClick(t);
                }
            }
        };

        v.findViewById(id).setOnClickListener(l);

        return l;
    }
    
    /**
     * Updates this Preference's state to coincide with the switch's state
     * (after the switch has changed state)
     * @param v the view witch contains the switch
     */
    private void updateSwitch(View v) {
        Switch s = (Switch) v.findViewById(R.id.switch_button);
        enabled = s.isChecked();

        v.findViewById(android.R.id.title).setEnabled(enabled);
        v.findViewById(android.R.id.summary).setEnabled(enabled);
        v.findViewById(R.id.edit).setEnabled(enabled);
    }

}

File: app/src/main/java/com/amaze/filemanager/ui/views/preference/PathSwitchPreference.kt
/*
 * Copyright (C) 2014-2021 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
 * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
 *
 * This file is part of Amaze File Manager.
 *
 * Amaze File Manager is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.amaze.filemanager.ui.views.preference

import android.content.Context
import android.view.View
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.amaze.filemanager.R

/** @author Emmanuel on 17/4/2017, at 22:22.
 */
class PathSwitchPreference(
    context: Context,
    private val onEdit: (PathSwitchPreference) -> Unit,
    private val onDelete: (PathSwitchPreference) -> Unit
) : Preference(context) {
    var lastItemClicked = -1
        private set

    init {
        widgetLayoutResource = R.layout.namepathswitch_preference
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        holder.itemView.let { view ->
            view.findViewById<View>(R.id.edit).setOnClickListener { onEdit(this) }
            view.findViewById<View>(R.id.delete).setOnClickListener { onDelete(this) }
            view.setOnClickListener(null)
        }

        // Keep this before things that need changing what's on screen
        super.onBindViewHolder(holder)
    }
}
